tests/net/pm/src/main.c

```diff
@@ -0,0 +1,189 @@
+		/* 00-00-5E-00-53-xx Documentation RFC 7042 */
+		ctx->mac_addr[0] = 0x00;
+		ctx->mac_addr[1] = 0x00;
+		ctx->mac_addr[2] = 0x5E;
+		ctx->mac_addr[3] = 0x00;
+		ctx->mac_addr[4] = 0x53;
+		ctx->mac_addr[5] = sys_rand32_get();
+	}
+
+	return ctx->mac_addr;
+}
+
+static void fake_dev_iface_init(struct net_if *iface)
+{
+	struct device *dev = net_if_get_device(iface);
+	struct fake_dev_context *ctx = dev->driver_data;
+	u8_t *mac = fake_dev_get_mac(ctx);
+
+	net_if_set_link_addr(iface, mac, 6, NET_LINK_ETHERNET);
+
+	ctx->iface = iface;
+}
+
+int fake_dev_init(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+struct fake_dev_context fake_dev_context_data;
+
+static struct dummy_api fake_dev_if_api = {
+	.iface_api.init = fake_dev_iface_init,
+	.send = fake_dev_send,
+};
+
+#define _ETH_L2_LAYER    DUMMY_L2
+#define _ETH_L2_CTX_TYPE NET_L2_GET_CTX_TYPE(DUMMY_L2)
+
+NET_DEVICE_INIT(fake_dev, "fake_dev",
+		fake_dev_init, fake_dev_pm_control,
+		&fake_dev_context_data, NULL,
+		CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		&fake_dev_if_api, _ETH_L2_LAYER, _ETH_L2_CTX_TYPE, 127);
+
+void test_setup(void)
+{
+	struct net_if *iface = net_if_get_default();
+	struct in_addr in4addr_my = { { { 192, 168, 0, 2 } } };
+	struct net_if_addr *ifaddr;
+
+	net_if_up(iface);
+
+	ifaddr = net_if_ipv4_addr_add(iface, &in4addr_my, NET_ADDR_MANUAL, 0);
+	zassert_not_null(ifaddr, "Could not add iface address");
+}
+
+void test_pm(void)
+{
+	struct net_if *iface = net_if_get_default();
+	struct device *dev = net_if_get_device(iface);
+	char data[] = "some data";
+	struct sockaddr_in addr4;
+	int sock;
+	int ret;
+
+	addr4.sin_family = AF_INET;
+	addr4.sin_port = htons(12345);
+	inet_pton(AF_INET, "192.168.0.1", &addr4.sin_addr);
+
+	sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	zassert_true(sock >= 0, "Could not open socket");
+
+	zassert_false(net_if_is_suspended(iface), "net iface is not suspended");
+
+	/* Let's send some data, it should go through */
+	ret = sendto(sock, data, ARRAY_SIZE(data), 0,
+		     (struct sockaddr *)&addr4, sizeof(struct sockaddr_in));
+	zassert_true(ret > 0, "Could not send data");
+
+	/* Let's make sure net stack's thread gets ran, or setting PM state
+	 * might return -EBUSY instead
+	 */
+	k_yield();
+
+	ret = device_set_power_state(dev, DEVICE_PM_SUSPEND_STATE,
+				     NULL, NULL);
+	zassert_true(ret == 0, "Could not set state");
+
+	zassert_true(net_if_is_suspended(iface), "net iface is not suspended");
+
+	/* Let's try to suspend it again, it should fail relevantly */
+	ret = device_set_power_state(dev, DEVICE_PM_SUSPEND_STATE,
+				     NULL, NULL);
+	zassert_true(ret == -EALREADY, "Could change state");
+
+	zassert_true(net_if_is_suspended(iface), "net iface is not suspended");
+
+	/* Let's send some data, it should fail relevantly */
+	ret = sendto(sock, data, ARRAY_SIZE(data), 0,
+		     (struct sockaddr *)&addr4, sizeof(struct sockaddr_in));
+	zassert_false(ret > 0, "Could send data");
```

Comment:
My brain always twists when reading these assert so at least for me this would be easier to read if it says `zassert_true(ret < 0, "Could send data");` especially when the other asserts are using the true variant.